Obsidian.Nbt/ModifiedUtf8.cs

```diff
@@ -0,0 +1,466 @@
+            source = ref Unsafe.Add(ref source, 1);
+            destination = ref Unsafe.Add(ref destination, 1);
+        }
+    }
+
+    private static void GetStringAsciiScalar(ReadOnlySpan<byte> bytes, ref char destination)
+    {
+        for (int i = 0; i < bytes.Length; i++)
+        {
+            destination = (char)bytes[i];
+            destination = ref Unsafe.Add(ref destination, 1);
+        }
+    }
+
+    private static void GetStringScalar(ReadOnlySpan<byte> bytes, ref char destination)
+    {
+        ref byte source = ref MemoryMarshal.GetReference(bytes);
+        ref byte sourceEnd = ref Unsafe.Add(ref source, bytes.Length);
+        while (Unsafe.IsAddressLessThan(ref source, ref sourceEnd))
+        {
+            int c = source;
+            int header = c >> 4;
+            if (header < 0b1000) // One byte
+            {
+                destination = (char)c;
+            }
+            else if (header < 0b1110) // Two bytes
+            {
+                c = (c & 0b0001_1111) << 6;
+
+                source = ref Unsafe.Add(ref source, 1);
+                c |= source & 0b0011_1111;
+
+                destination = (char)c;
+            }
+            else // Three bytes
+            {
+                c = (c & 0b0000_1111) << 12;
+
+                source = ref Unsafe.Add(ref source, 1);
+                c |= (source & 0b0011_1111) << 6;
+
+                source = ref Unsafe.Add(ref source, 1);
+                c |= source & 0b0011_1111;
+
+                destination = (char)c;
+            }
+            source = ref Unsafe.Add(ref source, 1);
+            destination = ref Unsafe.Add(ref destination, 1);
+        }
+    }
+
+    /// <summary>
+    /// Calculates the number of bytes produced by encoding the specified character span, unless the number of bytes is more than the encoding supports.
+    /// </summary>
+    /// <param name="chars">The span that contains the set of characters to encode.</param>
+    /// <param name="byteCount">The number of bytes produced by encoding the specified character span.</param>
+    /// <returns><c>true</c> if calculating the number of bytes was successful; otherwise <c>false</c></returns>
+    public static bool TryGetByteCount(ReadOnlySpan<char> chars, out int byteCount)
+    {
+        if (chars.Length > ushort.MaxValue) // Even for all-ASCII inputs, this produces > 65535 bytes
+        {
+            byteCount = default;
+            return false;
+        }
+
+        if (chars.IsEmpty)
+        {
+            byteCount = 0;
+            return true;
+        }
+
+        if (Avx2.IsSupported)
+        {
+            byteCount = GetByteCountAvx2(chars);
+            return byteCount <= ushort.MaxValue;
+        }
+        else
+        {
+            byteCount = GetByteCountScalar(chars);
+            return byteCount <= ushort.MaxValue;
+        }
+    }
+
+    private static int GetByteCountAvx2(ReadOnlySpan<char> chars)
+    {
+        Debug.Assert(Avx2.IsSupported);
```

Comment:
What happens when AVX2 isn't supported? Does this lose support for e.g. raspberry pis? Have you taken this into account?